File: packages/react-native-nitro-modules/ios/core/HybridObjectSpec.swift
//
//  HybridObjectSpec.swift
//  NitroModules
//
//  Created by Marc Rousavy on 23.07.24.
//

import Foundation

/**
 * A base protocol for all Swift-based Hybrid Objects.
 */
public protocol HybridObjectSpec: AnyObject {
  /**
   * Get the memory size of the Swift instance (plus any external heap allocations),
   * in bytes.
   *
   * Override this to allow tracking heap allocations such as buffers or images,
   * which will help the JS GC be more efficient in deleting large unused objects.
   *
   * @example
   * ```swift
   * var memorySize: Int {
   *   let imageSize = self.uiImage.bytesPerRow * self.uiImage.height
   *   return getSizeOf(self) + imageSize
   * }
   * ```
   */
  var memorySize: Int { get }
}

public extension HybridObjectSpec {
  /**
   * Get the memory size of the given instance.
   * This only accounts for stack allocated member variables,
   * not for externally allocated heap allocations like images or buffers.
   */
  func getSizeOf<T: AnyObject>(_ instance: T) -> Int {
    return malloc_size(Unmanaged.passUnretained(instance).toOpaque())
  }
}

File: packages/react-native-nitro-modules/android/src/main/java/com/margelo/nitro/HybridObjectSpec.kt
package com.margelo.nitro

import androidx.annotation.Keep
import com.facebook.jni.HybridData
import com.facebook.proguard.annotations.DoNotStrip

/**
 * A base protocol for all Kotlin-based Hybrid Objects.
 */
interface HybridObjectSpec {
    /**
     * Get the memory size of the Kotlin instance (plus any external heap allocations),
     * in bytes.
     *
     * Override this to allow tracking heap allocations such as buffers or images,
     * which will help the JS GC be more efficient in deleting large unused objects.
     *
     * @example
     * ```kotlin
     * val memorySize: ULong
     *   get() {
     *     val imageSize = this.bitmap.bytesPerRow * this.bitmap.height
     *     return getSizeOf(this) + imageSize
     *   }
     * ```
     */
    @DoNotStrip
    @Keep
    val memorySize: ULong

    /**
     * The C++-based state of the HybridObject.
     * This must be initialized in the Kotlin constructor via `initHybrid()`.
     */
    @DoNotStrip
    @Keep
    val mHybridData: HybridData

    /**
     * Initialize the C++ part of the HybridObject.
     * This must be called in the constructor, and the result must be assigned to `mHybridData`.
     */
    external fun initHybrid(): HybridData
}
